src/daemon/default_vm_image_vault.cpp

```diff
@@ -394,6 +393,24 @@ mp::VMImage mp::DefaultVMImageVault::fetch_image(const FetchType& fetch_type, co
 
         auto id = info.id.toStdString();
 
+        std::unique_lock<decltype(fetch_mutex)> lock{fetch_mutex};
+        auto it = in_progress_image_fetches.find(id);
+        if (it != in_progress_image_fetches.end())
+        {
+            auto future = it->second;
+            monitor(LaunchProgress::WAITING, -1);
+            lock.unlock();
+
+            auto prepared_image = future.result();
+
+            lock.lock();
+            prepared_image_records[id].last_accessed = std::chrono::system_clock::now();
+            auto vm_image = finalize_image_records(query, prepared_image);
+            lock.unlock();
```

Comment:
Yes, I tried that originally, but with how this function is laid out, it would be nearly impossible to use scoped locks.  Perhaps with breaking the function up in pieces, it would be easier, but that's for another day.

Reply:
Right, I really think it would be better to extract those pieces along with locking. It would make it much easier to understand and also maintain in the future.